OSDOCS#6055: Authenticating cert-manager Operator on GCP workload identity (#59895)
openshift/openshift-docs: modules/cert-manager-configure-cloud-credentials-gcp-sts.adoc

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-gcp-sts"]
= Configuring cloud credentials for the {cert-manager-operator} with GCP Workload Identity

> You can configure cloud credentials for the {cert-manager-operator} on the GCP Workload Identity cluster by using the `ccoctl` binary.

Reviewer: ```suggestion
In order to configure the cloud credentials for the {cert-manager-operator} on a GCP Workload Identity based cluster you need to use the `ccoctl` binary to generate the cloud credentials manually and apply it on the cluster.
```

.Prerequisites

* You must extract and prepare the `ccoctl` binary.
* The {cert-manager-operator} 1.11.1 or later is installed.

.Procedure

> . Create a directory to store the `CredentialsRequest` custom resource (CR) of the {cert-manager-operator} by running the following command:

Reviewer: Similar to https://github.com/openshift/openshift-docs/pull/59717#discussion_r1208926122
```suggestion
. Create a directory to store the `CredentialsRequest` custom resource (CR) by running the following command:
```

+
[source,terminal]
----
$ mkdir credentials-request
----

. Change the directory by running the following command:
+
[source,terminal]
----
$ cd credentials-request
----

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  generation: 1
  labels:
    app: cert-manager
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: cert-manager
    app.kubernetes.io/name: cert-manager
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

> . Use the `ccoctl` tool to process `CredentialsRequest` objects of the {cert-manager-operator} by running the following command:

Reviewer: ```suggestion
. Use the `ccoctl` tool to process `CredentialsRequest` object(s) by running the following command:
```
Similar to https://github.com/openshift/openshift-docs/pull/59717#discussion_r1208926669

+
[source,terminal]
----
$ ccoctl gcp create-service-accounts \
    --name <name> --output-dir=<path_to_output_dir> \
    --credentials-requests-dir=<path_to_credrequests_dir> \
    --workload-identity-pool <workload_identity_pool> \
    --workload-identity-provider <workload_identity_provider> \
    --project <project_name>
----

. Apply the secrets generated in the manifests directory of your cluster by running the following command:
+
[source,terminal]
----
$ ls manifests/*-credentials.yaml | xargs -I{} oc apply -f {}
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
----

> .Verification

Reviewer: Can you update this section to verify from the pod instead of the deployment? I think pod makes more sense (and that approach is more specific and accurate).
Exactly similar to what you did in https://github.com/openshift/openshift-docs/pull/59907 (copy paste is all we need).

```
.Verification
. Get the name of a cert-manager controller pod by running the following command:
+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                          READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt  1/1     Running   0          15m39s
----
. Verify that AWS credentials are updated by running the following command:
+
[source,terminal]
----
$ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'
----
+
.Example output
[source,terminal]
----
- name: cloud-credentials
..........
````

* Verify that the secret is updated by running the following command:
+
[source,terminal]
----
$ oc get deployment cert-manager -n cert-manager -o yaml
----
+
.Example output
[source,terminal]
----

> - name: cloud-credentials

Reviewer: Minor indentation nit (to make output look more realistic):
```suggestion
---
       - name: cloud-credentials
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-gcp-sts"]
= Configuring cloud credentials for the {cert-manager-operator} with GCP Workload Identity

To configure the cloud credentials for the {cert-manager-operator} on a GCP Workload Identity cluster, you must use the `ccoctl` binary to generate the cloud credentials manually and apply it on the cluster.

.Prerequisites

> * You must extract and prepare the `ccoctl` binary.

Reviewer 2: In our cert-manager document, you should share a link (or links) for OpenShift `GCP Workload Identity cluster` and for `ccoctl` extraction. Take 4.13 for example, all are in https://docs.openshift.com/container-platform/4.13/authentication/managing_cloud_provider_credentials/cco-mode-gcp-workload-identity.html , specifically [`ccoctl` extraction](https://docs.openshift.com/container-platform/4.13/authentication/managing_cloud_provider_credentials/cco-mode-gcp-workload-identity.html#cco-ccoctl-configuring_cco-mode-gcp-workload-identity) is there.

Author: Hi, @xingxingxia I added the link for the "Configuring the Cloud Credential Operator utility" section in the additional resources section. Do we also need the link for the "Installing an OpenShift Container Platform cluster configured for manual mode with GCP Workload Identity" section in the additional resources section?

Author: Ah, I suppose yes!

* The {cert-manager-operator} 1.11.1 or later is installed.

> 

Reviewer 2: @swghosh , @xenolinux , one more prerequisite is needed: the OpenShift cluster is configured with https://docs.openshift.com/container-platform/4.13/authentication/managing_cloud_provider_credentials/cco-mode-gcp-workload-identity.html

.Procedure

. Create a directory to store the `CredentialsRequest` custom resource (CR) by running the following command:
+
[source,terminal]
----
$ mkdir credentials-request
----

. Change the directory by running the following command:
+
[source,terminal]
----
$ cd credentials-request
----

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  generation: 1
  labels:
    app: cert-manager
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: cert-manager

>     app.kubernetes.io/name: cert-manager

Reviewer 2: These generation and labels are not needed.

  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:

>     - roles/dns.admin

Reviewer 2: It is better to add a note for the role(s) used, like https://cert-manager.io/docs/configuration/acme/dns01/google/ noted:
```
Note: The use of the dns.admin role in this example role is 
for convenience. If you want to ensure cert-manager runs 
under a least privilege service account, you will need to 
create a custom role with the following permissions:
  dns.resourceRecordSets.*
  dns.changes.*
  dns.managedZones.list
```

  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Use the `ccoctl` tool to process `CredentialsRequest` objects by running the following command:
+
[source,terminal]
----

> $ ccoctl gcp create-service-accounts \
>     --name <name> --output-dir=<path_to_output_dir> \
>     --credentials-requests-dir=<path_to_credrequests_dir> \
>     --workload-identity-pool <workload_identity_pool> \
>     --workload-identity-provider <workload_identity_provider> \
>     --project <project_name>

Reviewer 2: ```suggestion
$ ccoctl gcp create-service-accounts \
    --name <user_defined_name> --output-dir=<path_to_output_dir> \
    --credentials-requests-dir=<path_to_credrequests_dir> \
    --workload-identity-pool <workload_identity_pool> \
    --workload-identity-provider <workload_identity_provider> \
    --project <gcp_project_id>
```

----

. Apply the secrets generated in the manifests directory of your cluster by running the following command:
+
[source,terminal]
----

> $ ls manifests/*-credentials.yaml | xargs -I{} oc apply -f {}

Reviewer 2: This is incorrect. The arg should corresponds to above `--output-dir=<path_to_output_dir>`. Ref: https://docs.openshift.com/container-platform/4.13/updating/preparing-manual-creds-update.html#cco-ccoctl-upgrading_preparing-manual-creds-update
```suggestion
$ ls <path_to_output_dir>/manifests/*-credentials.yaml | xargs -I{} oc apply -f {}
```

----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----

> $ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'

Reviewer 2: ```suggestion
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type=merge -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"gcp-credentials"}]}}}'
```

----

.Verification

> . Get the name of a cert-manager controller pod by running the following command:

Reviewer 2: The pod will be redeployed.
```suggestion
. Get the name of the redeployed cert-manager controller pod by running the following command:
```

+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                          READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt  1/1     Running   0          15m39s
----

. Verify that GCP credentials are updated by running the following command:
+
[source,terminal]
----

> $ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'

Reviewer 2: This will be incorrect. `-o=jsonpath='{.spec.volumes}'` will print JSON format like, but your Example output is YAML format.

----
+
.Example output
[source,terminal]
----

> - name: cloud-credentials
>     secret:
>       secretName: gcp-creds

Reviewer 2: You should mention the pod is updated with below GCP workload identity credential volumes mounted under below paths:
`oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o yaml` example output:
```suggestion
spec:
  containers:
  - args:
    ...
    volumeMounts:
    - mountPath: /var/run/secrets/openshift/serviceaccount
      name: bound-sa-token
      ...
    - mountPath: /.config/gcloud
      name: cloud-credentials
  ...
  volumes:
  - name: bound-sa-token
    projected:
      ...
      sources:
      - serviceAccountToken:
          audience: openshift
          ...
          path: token
  - name: cloud-credentials
    secret:
      defaultMode: 420
      items:
      - key: service_account.json
        path: application_default_credentials.json
      secretName: cloud-credentials
```
https://issues.redhat.com/browse/OCPBUGS-9998?focusedId=22351582&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22351582 verification has details if needed for reference.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-gcp-sts_{context}"]
= Configuring cloud credentials for the {cert-manager-operator} with GCP Workload Identity

To configure the cloud credentials for the {cert-manager-operator} on a GCP Workload Identity cluster, you must generate the cloud credentials manually by using the `ccoctl` binary, and apply it on the cluster.

.Prerequisites

* You must extract and prepare the `ccoctl` binary.
* The {cert-manager-operator} 1.11.1 or later is installed.
* You have configured an {product-title} cluster with GCP Workload Identity by using the Cloud Credential Operator in a manual mode.

.Procedure

. Create a directory to store a `CredentialsRequest` resource YAML file by running the following command:
+
[source,terminal]
----
$ mkdir credentials-request
----

> . Create a `CredentialsRequest` resource YAML file under the `credentials-request` directory, such as, `sample-credential-request.yaml` by applying the following yaml:

Reviewer 3: This step can be clarified; suggest:

```
. In the `credentials-request` directory, create a YAML file that contains the following `CredentialsRequest` manifest:
```

I don't think the example file name (`sample-credential-request.yaml`) is necessary because it doesn't come up in the rest of the procedure.

+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----
+
[NOTE]
====
The `dns.admin` role provides admin privileges to the service account for managing Google Cloud DNS resources. To ensure cert-manager runs with the service account that has the least privilege, you must create a custom role with the following permissions:

* `dns.resourceRecordSets.*`
* `dns.changes.*`
* `dns.managedZones.list`
====

> . Use the `ccoctl` tool to process `CredentialsRequest` objects by running the following command:

Reviewer 3: Future improvement suggestion: add an example of this command with all of the user-replaceable values substituted with example values.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-gcp-sts"]
= Configuring cloud credentials for the {cert-manager-operator} with GCP Workload Identity

To configure the cloud credentials for the {cert-manager-operator} on a GCP Workload Identity cluster, you must use the `ccoctl` binary to generate the cloud credentials manually and apply it on the cluster.

.Prerequisites

* You must extract and prepare the `ccoctl` binary.
* The {cert-manager-operator} 1.11.1 or later is installed.
* You have configured an {product-title} cluster by using a manual mode with GCP Workload Identity.

.Procedure

> . Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:

Reviewer 2: Sorry forgot one more thing, you need add below step before line 19:
```
. Create a directory to store the `CredentialsRequest` custom resource (CR) by running the following command:
+
[source,terminal]
----
$ mkdir credentials-request

```

And update
```
Create a `CredentialsRequest` resource YAML file
```
to be:
```
Create a `CredentialsRequest` resource YAML file under `credentials-request` directory
```

Author: Addressed

+
[source,terminal]
----
$ mkdir credentials-request
----

. Create a `CredentialsRequest` resource YAML file under the `credentials-request` directory, such as, `sample-credential-request.yaml` by applying the following yaml:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: GCPProviderSpec
    predefinedRoles:
    - roles/dns.admin
  secretRef:
    name: gcp-credentials
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----
+
[NOTE]
====

> The use of the `dns.admin` role in the `sample-credential-request.yaml` file is for convenience. To ensure cert-manager runs

Author: Updated the note

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

> 

Reviewer 2: It is NOT needed to manually create the `CredentialsRequest` resource under CCO manual mode. Following ccoctl does process it. Remove above lines.

+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                          READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt  1/1     Running   0          15m39s
----

. Verify that the cert-manager controller pod is updated with GCP workload identity credential volumes mounted under the path specified in `mountPath` by running the following command:
+
[source,terminal]
----
$ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o yaml
----
+
.Example output
[source,terminal]
----
spec:
  containers:
  - args:
    ...
    volumeMounts:
    - mountPath: /var/run/secrets/openshift/serviceaccount
      name: bound-sa-token
      ...
    - mountPath: /.config/gcloud
      name: cloud-credentials
  ...
  volumes:
  - name: bound-sa-token
    projected:
      ...
      sources:
      - serviceAccountToken:
          audience: openshift
          ...
          path: token
  - name: cloud-credentials
    secret:
      ...
      items:
      - key: service_account.json
        path: application_default_credentials.json

>       secretName: cloud-credentials

Reviewer 2: Above sample-credential-request.yaml secretRef name is `gcp-credentials`, secretName means that.
```suggestion
      secretName: gcp-credentials
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-gcp.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-gcp-sts"]
= Preparing cloud credentials for the {cert-manager-operator} with GCP Workload Identity

You can prepare cloud credentials for the {cert-manager-operator} on GCP workload identity by using the `ccoctl` binary.

.Prerequisites

* You must extract and prepare the `ccoctl` binary.

.Procedure

. The {cert-manager-operator} creates the `CredentialsRequest` object in the `openshift-cloud-credential-operator` namespace. Get the role by running the following command:
+
[source,terminal]
----
$ oc get credentialsrequest -n openshift-cloud-credential-operator -o yaml
----
+
.Example output
[source,terminal]
----
output
----

. Annotate the service account by running the following command:
+
[source,terminal]
----

> $ oc annotate sa/cert-manager -n cert-manager to-do-value

Reviewer: Do not need this annotate in GCP workload identity, the other steps should suffice.
